docs/development-guide.md

```diff
@@ -0,0 +1,192 @@
+
+To build ceph-csi-operator in a container:
+
+```console
+make docker-build
+```
+
+The built binary will be present under `bin/` directory.
+
+### Running ceph-csi-operator tests
+
+Once the changes to the sources compile, it is good practise to run the tests
+that validate the style and other basics of the source code. Execute the unit
+tests (in the `*_test.go` files) and check the formatting of YAML files,
+MarkDown documents and shell scripts:
+
+```console
+make test
+```
+
+In addition to running tests locally, each Pull Request that is created will
+trigger Continuous Integration tests, but
+also additional functionality tests that are defined under the `e2e/`
+directory.
+
+### Code contribution workflow
+
+ceph-csi-operator repository currently follows GitHub's
+[Fork & Pull] (<https://help.github.com/articles/about-pull-requests/>) workflow
+for code contributions.
+
+Please read the [coding guidelines](coding.md) document before submitting a PR.
+
+#### Certificate of Origin
+
+By contributing to this project you agree to the Developer Certificate of
+Origin (DCO). This document was created by the Linux Kernel community and is a
+simple statement that you, as a contributor, have the legal right to make the
+contribution. See the [DCO](DCO) file for details.
+
+Contributors sign-off that they adhere to these requirements by adding a
+Signed-off-by line to commit messages. For example:
+
+```
+subsystem: This is my commit message
+
+More details on what this commit does
+
+Signed-off-by: Random J Developer <random@developer.example.org>
+```
+
+If you have already made a commit and forgot to include the sign-off, you can
+amend your last commit to add the sign-off with the following command, which
+can then be force pushed.
+
+```console
+git commit --amend -s
+```
+
+We use a [DCO bot](https://github.com/apps/dco) to enforce the DCO on each pull
+request and branch commits.
+
+#### Commit Messages
+
+We follow a rough convention for commit messages that is designed to answer two
+questions: what changed and why? The subject line should feature the what and
+the body of the commit should describe the why.
+
+```
+controller: fix bug in configmap
+
+fix clusterID bug in the configmap where the clusterID is
+not set to the expected value.
+
+Signed-off-by: Random J Developer <random@developer.example.org>
+```
+
+The format can be described more formally as follows:
+
+```
+<component>: <subject of the change>
+<BLANK LINE>
+<paragraph(s) with reason/description>
+<BLANK LINE>
+<signed-off-by>
+```
+
+The first line is the subject and should be no longer than 70 characters, the
+second line is always blank, and other lines should be wrapped at 80 characters.
+This allows the message to be easier to read on GitHub as well as in various
+git tools.
+
+Here is a short guide on how to work on a new patch.  In this example, we will
+work on a patch called *hellopatch*:
+
+```console
+git checkout main
+git pull
+git checkout -b hellopatch
```

Comment:
Why use git pull? git pull creates a local merge request on the local main which is usually not the desired result. 
The alternative to sync your local main with upstream main is:
```
git fetch upstream main:main
git checkout main 
git push 
```
This will guarantee that your fork's main does not have any artifacts/merge commits that are not in upstream main

Then you can run
```
git checkout -b path_branch_name
```